unified_planning/engines/temporal_simulator.py

```diff
@@ -0,0 +1,941 @@
+        the `grounded Action` itself, and adds the corresponding `List of Events` to this `Simulator`. If the
+        corresponding `Events` were already created, the same value is returned and no new `Events` are created.
+
+        :param original_action: The `Action` of the :class:`~unified_planning.model.Problem` grounded with the given `params`.
+        :param params: The expressions used to ground the `original_action`.
+        :param grounded_action: The grounded action, result of the `original_action` grounded with the given `parameters`.
+        :return: The retrieved or created `List of Events` corresponding to the `grounded_action`.
+        """
+        if isinstance(original_action, InstantaneousAction):
+            assert isinstance(grounded_action, InstantaneousAction)
+            return [
+                TemporalEvent(
+                    TemporalEventKind.INSTANTANEOUS_ACTION,
+                    StartTiming(),
+                    True,
+                    None,
+                    grounded_action.preconditions,
+                    grounded_action.effects,
+                    grounded_action.simulated_effect,
+                )
+            ]
+        elif isinstance(original_action, DurativeAction):
+            if (
+                grounded_action is None
+            ):  # The grounded action is meaningless, no event associated
+                event_list: List["Event"] = []
+            else:
+                assert isinstance(grounded_action, DurativeAction)
+                event_list = break_action_or_problem_in_event_list(
+                    grounded_action.effects,
+                    grounded_action.conditions,
+                    grounded_action.simulated_effects,
+                    duration,
+                    grounded_action.env,
+                    is_global=False,
+                )
+            return event_list
+        else:
+            raise NotImplementedError(
+                "The TemporalSimulator currently supports only InstantaneousActions and DurativeActions."
+            )
+
+    def _expand_event(
+        self,
+        event: "TemporalEvent",
+        stn: DeltaSimpleTemporalNetwork,
+        running_events: List[List[Event]],
+        durative_conditions: List[List[FNode]],
+        last_event: "TemporalEvent",
+    ):
+        """IMPORTANT NOTE: This function modifies the data structures given as parameters."""
+        insert_interval(stn, last_event, event, left_bound=Fraction(0))
+        if event.kind == TemporalEventKind.START_ACTION:
+            committed_events: List[Event] = []
+            ce = event.committed_events
+            assert (
+                ce is not None
+            ), "Error, start Action event can't have None committed events"
+            for committed_event in ce:
+                if any(committed_event in el for el in running_events):
```

Comment:
Here I am thinking about a better way to find if an Event was already submitted in this "states chain".

The best idea I had so far is to hold a set in the state with all the events already applied. It might also be a `set[int]`, with only the events id's, if it helps.